src/Microsoft.OData.Edm/Vocabularies/Annotations/EdmDirectValueAnnotationsManager.cs

```diff
@@ -173,9 +173,13 @@ private static void SetAnnotation(IEnumerable<IEdmDirectValueAnnotation> immutab
             bool needTombstone = false;
             if (immutableAnnotations != null)
             {
-                if (immutableAnnotations.Any(existingAnnotation => existingAnnotation.NamespaceUri == namespaceName && existingAnnotation.Name == localName))
+                foreach (var existingAnnotation in immutableAnnotations)

```

Comment:
I don't immediately see how this code is leading to fewer allocations.  Can you please explain.

Reply:
Every time we invoke Any, it allocates a lambda. This is normally meaningless, when not present in the hot path. Since this is present in AGS hot path, it is costing 1% of allocations just in this lambda ...This is a lot.

Not using Linq.Any(), is not only good for memory but also for CPU as you can see in this benchmark:

![image](https://user-images.githubusercontent.com/24846248/128551846-9cf2d500-4992-412e-a003-4d04f6d53c96.png)

Reply:
I strive for readability and less code verbosity, as much as possible. But when it comes to choose between readability versus performance and COGs, I think we all agree with the choice.

IMHO, I don't think this code hurts readability that much, however it is more verbose I agree. But that is a non-goal when addressing performance issues of code that runs at scale.

In fact, we should encourage developers to stop using Linq for code that runs at scale - not the other way around.

Reply:
Thanks JP. And I fully support your change. Want to learn more about the tool you used to analyze this.

But just for the record (and not reverting on my support for this)), no in the absolut way you state "choose between readability versus performance" I do actually disagree. It is comparing the easily measurable cost of performance to the much harder to measure and long term changing cost of maintenance. If we would (could) measure this, the winner is not that obvious anymore.

Reply:
We discourage use of `var`. You can use the type.